File: liankao/25.01.03/stole/gpt.cpp
#include <bits/stdc++.h>
using namespace std;

// Input and output files
ifstream inputFile("stole.in");
ofstream outputFile("stole.out");

class FileHandler {
public:
    ~FileHandler() {
        inputFile.close();
        outputFile.close();
    }
} fileHandler;

using ull = unsigned long long;

class MathUtils {
private:
    const int mod;

public:
    constexpr MathUtils(int p) : mod(p) {}

    constexpr ull modPow(ull base, ull exp) const {
        ull result = 1;
        while (exp) {
            if (exp & 1) result = (result * base) % mod;
            base = (base * base) % mod;
            exp >>= 1;
        }
        return result;
    }

    constexpr ull modInverse(ull x) const {
        return modPow(x, mod - 2);
    }
};

class LagrangeInterpolation {
public:
    ull interpolate(const vector<ull>& y, const vector<ull>& x, ull evalX, int mod) const {
        int n = y.size();
        ull result = 0;
        MathUtils math(mod);

        for (int i = 0; i < n; ++i) {
            ull term = y[i];
            for (int j = 0; j < n; ++j) {
                if (i != j) {
                    term = term * (evalX + mod - x[j]) % mod * math.modInverse((x[i] + mod - x[j]) % mod) % mod;
                }
            }
            result = (result + term) % mod;
        }
        return result;
    }
};

int main() {
    ios::sync_with_stdio(false);
    inputFile.tie(nullptr);

    int n, r, mod;
    inputFile >> n >> r >> mod;

    MathUtils math(mod);

    // Precompute factorials and modular inverses
    vector<ull> factorial(n + 7, 1), invFactorial(n + 7);
    for (int i = 1; i < n + 7; ++i) {
        factorial[i] = factorial[i - 1] * i % mod;
    }
    for (int i = 0; i < n + 7; ++i) {
        invFactorial[i] = math.modInverse(factorial[i]);
    }

    auto combination = [&](int a, int b) -> ull {
        if (a < b) return 0;
        return factorial[a] * invFactorial[b] % mod * invFactorial[a - b] % mod;
    };

    const int m = min(n * (n - 1) / 2, r) + 7;

    // DP table for computing f
    vector<vector<ull>> f(n + 1, vector<ull>(m + 1));
    fill(f[1].begin() + 1, f[1].end(), 1);

    for (int w = 1; w <= m; ++w) {
        vector<ull> powers(n * n + 7, 1);
        for (int i = 1; i < n * n + 7; ++i) {
            powers[i] = powers[i - 1] * (mod + r - w + 2) % mod;
        }

        vector<vector<ull>> g(n + 1, vector<ull>(n + 1));
        g[1][0] = 1;

        for (int a = 2; a <= n; ++a) {
            f[a][w] = f[a][w - 1];
            g[a][0] = f[a][w - 1];

            for (int k = 1; k < n; ++k) {
                for (int b = 1; b < a; ++b) {
                    for (int x = 0; x < k; ++x) {
                        ull temp = g[b][x] * g[a - b][k - x - 1] % mod;
                        temp = temp * combination(a, 2) % mod;
                        temp = temp * combination(a - 2, b - 1) % mod;
                        temp = temp * powers[b * (a - b) - 1] % mod;
                        g[a][k] = (g[a][k] + temp) % mod;
                    }
                }
                g[a][k] = g[a][k] * math.modInverse(k) % mod;
                f[a][w] = (f[a][w] + g[a][k]) % mod;
            }
        }
    }

    vector<ull> fValues(m), positions(m);
    iota(positions.begin(), positions.end(), 1);
    for (int i = 0; i < m; ++i) {
        fValues[i] = f[n][i + 1];
    }

    LagrangeInterpolation lagrange;
    outputFile << lagrange.interpolate(fValues, positions, r, mod) << '\n';

    return 0;
}

File: liankao/25.01.03/stole/gpt.py
import sys
from functools import lru_cache

class FileIO:
    def __init__(self, input_file, output_file):
        self.input_file = input_file
        self.output_file = output_file
        self.inf = open(input_file, "r")
        self.ouf = open(output_file, "w")

    def read(self):
        return self.inf.read()

    def write(self, data):
        self.ouf.write(data)

    def close(self):
        self.inf.close()
        self.ouf.close()

# Handle input/output using FileIO
file_io = FileIO("stole.in", "stole.out")
sys.stdin = file_io.inf
sys.stdout = file_io.ouf

class MathBase:
    def __init__(self, p):
        self.p = p

    def qpow(self, a, b):
        res = 1
        while b:
            if b & 1:
                res = res * a % self.p
            a = a * a % self.p
            b >>= 1
        return res

    def inv(self, x):
        return self.qpow(x, self.p - 2)

def main():
    input = sys.stdin.read
    data = input().split()
    n, r, p = map(int, data)
    
    math = MathBase(p)
    
    # Precompute factorials and their modular inverses
    fct = [1] * (n + 7)
    for i in range(1, n + 7):
        fct[i] = fct[i - 1] * i % p
    
    ifct = [math.inv(x) for x in fct]
    
    def C(a, b):
        if a < b:
            return 0
        return fct[a] * ifct[b] % p * ifct[a - b] % p

    m = n + 7
    f = [[0] * (m + 1) for _ in range(n + 1)]
    for w in range(1, m + 1):
        pw = [1] * (n * n + 7)
        for i in range(1, n * n + 7):
            pw[i] = pw[i - 1] * (m - w + 2) % p

        g = [[0] * (n + 1) for _ in range(n + 1)]
        g[1][0] = 1
        
        for a in range(2, n + 1):
            f[a][w] = f[a][w - 1]
            g[a][0] = f[a][w - 1]
            
            for k in range(1, n):
                for b in range(1, a):
                    for x in range(0, k):
                        g[a][k] = (g[a][k] +
                                   g[b][x] * g[a - b][k - x - 1] % p *
                                   C(a, 2) % p * C(a - 2, b - 1) % p *
                                   pw[b * (a - b) - 1]) % p
                
                g[a][k] = g[a][k] * math.inv(k) % p
                f[a][w] = (f[a][w] + g[a][k]) % p

    print(f[n][r])

if __name__ == "__main__":
    main()
    file_io.close()
